### PR TITLE
Add support for (DevkitPro)libnds

### PR DESCRIPTION
**Pull request by buchstabenwurst**
# Description

Added support for libnds from DevkitPro.

# Testing

How did you test?

With the configuration as written in ./IDE/NDS/README.md
 - Real Hardware
 - MelonDS Emulator

# Checklist

 - [X] added tests
 - [ ] updated/added doxygen
 - [X] updated appropriate READMEs
 - [X] Updated manual and documentation

### CODE REVIEW

**Comment by embhorn** [comment]
Hi @buchstabenwurst 

Thanks for your proposed changes. We require having a contributor agreement on file to accept changes from outside sources. I can help you get started if you email support@wolfssl.com and mention this PR.

Kind regards,
Eric - wolfSSL Support

**Comment by dgarske** [comment]
Hi @buchstabenwurst , your contributor agreement has been approved. Okay to test.

File: wolfcrypt/test/test.c [hunk 1]
```diff
@@ -2457,6 +2464,13 @@ options: [-s max_relative_stack_bytes] [-m max_relative_heap_memory_bytes]\n\
         VIDEO_WaitVSync();
         if(rmode->viTVMode&VI_NON_INTERLACE) VIDEO_WaitVSync();
 #endif
+#ifdef NDS
+        // Init Console output
```
Comment (dgarske) [inline wolfcrypt/test/test.c:2468]:
```
C++-style comments:
./wolfcrypt/test/test.c:2477:        // Init Console output
./wolfcrypt/test/test.c:2480:        // Init the Filesystem
./wolfcrypt/test/test.c:2529:        // in Nintendo DS returning from main shuts down the Device without letting you see the Results.
```

File: wolfcrypt/test/test.c [hunk 2]
```diff
@@ -426,6 +426,13 @@ const byte const_byte_array[] = "A+Gd\0\0\0";
 #ifdef DEVKITPRO
     #include <wiiuse/wpad.h>
 #endif
+#ifdef NDS
```
Comment (dgarske) [inline wolfcrypt/test/test.c:429]: Is `NDS` a compiler pre-processor macro that gets defined?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/test/test.c (hunk 2 above)
Comment (buchstabenwurst) [reply wolfcrypt/test/test.c:429]: It gets manually defined as descibed in ./IDE/NDS/README.md .
I'm sorry if it's not the correct way to do it, but I found it as the best way to seperate the parts for cross compilation.

Comment (dgarske) [reply wolfcrypt/test/test.c:429]: I see. It's not specific enough. Maybe change to WOLFSSL_NDS. Did you see the existing `DEVKITPRO` build option? Curious if you tried with that or maybe there is a way to expand that existing one?

Comment (buchstabenwurst) [reply wolfcrypt/test/test.c:429]: I will change that, as for the existing `DEVKITPRO` build option, it seems to be for the Wii and is incompatible. at least for displaying the tests.

Comment (buchstabenwurst) [reply wolfcrypt/test/test.c:429]: Shuld i really change it? For what i understand `WOLFSSL_xxx` seems to be for features, and just `xxx` for diffrent plattforms, as for the existing `DEVKITPRO` build option, it seems to be for the Wii and is incompatible. at least for displaying the tests.

Comment (dgarske) [reply wolfcrypt/test/test.c:429]: You can keep it as NDS for now. Thank you

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (buchstabenwurst) [reply wolfcrypt/test/test.c:2468]: Changed it

File: wolfssl/wolfcrypt/settings.h [hunk 3]
```diff
@@ -469,6 +472,23 @@
...
+#ifdef NDS
+    #include <stddef.h>
+    #define SIZEOF_LONG_LONG 8
+    #define socklen_t int
+    #define IPPROTO_UDP 17
+    #define IPPROTO_TCP 6
+
+    /* Libnds doesn't include sys/uio.h.  */
+    /* Structure for scatter/gather I/O.  */
+    struct iovec
```
Comment (dgarske) [inline wolfssl/wolfcrypt/settings.h:485]: Note: This can also be resolved by `#define NO_WRITEV`.

**Review by dgarske — APPROVED** [review]

File: wolfssl/wolfcrypt/settings.h (hunk 3 above)
Comment (buchstabenwurst) [reply wolfssl/wolfcrypt/settings.h]: Thanks, I changed it.

**Review by dgarske — APPROVED** [review]

**Comment by JacobBarthelmeh** [comment]
Retest this please Jenkins, the fail reason was not captured before the logs were erased.

**Comment by JacobBarthelmeh** [comment]
Errors reported with CI are not related to these code changes. Thanks for the contribution.

**Review by JacobBarthelmeh — APPROVED** [review]

**Comment by jackctj117** [comment]
@buchstabenwurst how were you able to read in the certs for the wolfcrypt test?

**Comment by buchstabenwurst** [comment]
> @buchstabenwurst how were you able to read in the certs for the wolfcrypt test?

I copied the `certs` directory to `_nds/certs` on the sd card of the ds.

**Comment by jackctj117** [comment]
> > @buchstabenwurst how were you able to read in the certs for the wolfcrypt test?
> 
> I copied the `certs` directory to `_nds/certs` on the sd card of the ds.

I was able to get it to work but I had to remove `fat:/` from `#define CERT_PREFIX "fat:/_nds/"` in order for MelonDS to read in the certs. Will this change the functionality when using physical hardware?

**Comment by buchstabenwurst** [comment]
> > > @buchstabenwurst how were you able to read in the certs for the wolfcrypt test?
> > 
> > 
> > I copied the `certs` directory to `_nds/certs` on the sd card of the ds.
> 
> I was able to get it to work but I had to remove `fat:/` from `#define CERT_PREFIX "fat:/_nds/"` in order for MelonDS to read in the certs. Will this change the functionality when using physical hardware?

This will work the same, the `fat:` is for compatability when using fat and nitrofs together.

**Pull request merged**